### PR TITLE
Configuring Console Commands via attributes

### PR DESCRIPTION
**Pull request by msmakouz**
| Q             | A
| ------------- | ---
| Bugfix?       | ❌
| Breaks BC?    | ❌ 
| New feature?  | ✔️

## Configuring Console Commands via attributes

Added the ability to configure console commands using attributes.

```php
namespace App\Api\Cli\Command;

use Spiral\Console\Attribute\Argument;
use Spiral\Console\Attribute\AsCommand;
use Spiral\Console\Attribute\Option;
use Spiral\Console\Attribute\Question;
use Spiral\Console\Command;
use Symfony\Component\Console\Input\InputOption;

#[AsCommand(name: 'app:create:user', description: 'Creates a user with the given data')]
final class CreateUser extends Command
{
    #[Argument]
    #[Question(question: 'Provide the User Email')]
    private string $email;

    #[Argument]
    #[Question(question: 'Provide the User password')]
    private string $password;

    #[Argument]
    #[Question(question: 'Provide the User first name')]
    private string $firstName;

    #[Argument]
    #[Question(question: 'Provide the User last name')]
    private string $lastName;

    #[Option(mode: InputOption::VALUE_NONE)]
    private bool $active;

    public function __invoke(): int
    {
        // ...

        return self::SUCCESS;
    }
}
```

All properties will be filled from the console input automatically and will be available with the required values in the **__invoke** method.

## Prompt for missing required arguments

Added the ability to prompt for missing required arguments. If the required `arguments` are not passed when calling the console command, a question will be asked with the ability to type the missing arguments instead of displaying an error.

Using the `Spiral\Console\Attribute\Question` attribute, can specify question texts for arguments.

- The attribute can be added to a class, with the required parameter the name of the argument and the text of the question.

```php
#[AsCommand(name: 'app:create:user', description: 'Creates a user with the given data')]
#[Question(question: 'Provide the User Email', argument: 'email')]
#[Question(question: 'Provide the User password', argument: 'password')]
#[Question(question: 'Provide the User first name', argument: 'firstName')]
final class CreateUser extends Command
{
    public function __invoke(): int
    {
        // ...

        return self::SUCCESS;
    }
}
```
- The attribute can be added to a property.

```php
#[AsCommand(name: 'app:create:user', description: 'Creates a user with the given data')]
final class CreateUser extends Command
{
    #[Argument]
    #[Question(question: 'Provide the User Email')]
    private string $email;

    #[Argument]
    #[Question(question: 'Provide the User password')]
    private string $password;

    #[Argument]
    #[Question(question: 'Provide the User first name')]
    private string $firstName;

    #[Argument]
    #[Question(question: 'Provide the User last name')]
    private string $lastName;

    #[Option(mode: InputOption::VALUE_NONE)]
    private bool $active;

    public function __invoke(): int
    {
        // ...

        return self::SUCCESS;
    }
}
```

### Disabling 

For disabling this functionality, need to override the method `interact` in the console command.

```php
protected function interact(InputInterface $input, OutputInterface $output): void
{
}
```

### CODE REVIEW

**Comment by butschster** [comment]
@msmakouz Add support for Symfony AsCommand attribue

File: src/Console/src/Configurator/Configurator.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+declare(strict_types=1);
+
+namespace Spiral\Console\Configurator;
+
+use Spiral\Console\Command;
+
+/**
+ * @internal
+ */
+final class Configurator implements ConfiguratorInterface
```
Comment (butschster) [inline src/Console/src/Configurator/Configurator.php:12]: Could you get rid of the interface here? Is it necessary?

**Review by butschster — CHANGES_REQUESTED** [review]

File: src/Console/src/Configurator/Attribute/Parser.php [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+final class Parser
+{
+    public function __construct(
+        private readonly ReaderInterface $reader = new AttributeReader()
+    ) {
+    }
+
+    public function hasCommandAttribute(\ReflectionClass $reflection): bool
+    {
+        return $this->reader->firstClassMetadata($reflection, AsCommand::class) !== null;
```
[the remark this answers is not in the document]
Comment (msmakouz) [reply src/Console/src/Configurator/Attribute/Parser.php]: Added

File: src/Console/src/Configurator/ConfiguratorInterface.php [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+namespace Spiral\Console\Configurator;
+
+use Spiral\Console\Command;
+
+/**
+ * @internal
+ */
+interface ConfiguratorInterface
+{
+    public function canConfigure(Command $command, \ReflectionClass $reflection): bool;
```
[the remark this answers is not in the document]
Comment (msmakouz) [reply src/Console/src/Configurator/ConfiguratorInterface.php:19]: Added phpdoc

File: src/Console/src/Configurator/ConfiguratorInterface.php [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+    public function configure(Command $command, \ReflectionClass $reflection): void;
```
[the remark this answers is not in the document]
Comment (msmakouz) [reply src/Console/src/Configurator/ConfiguratorInterface.php:26]: Added phpdoc

File: src/Console/src/Configurator/Configurator.php (hunk 1 above)
Comment (msmakouz) [reply src/Console/src/Configurator/Configurator.php]: Removed interface implementation here

File: src/Console/src/Configurator/Result.php [hunk 5]
```diff
@@ -2,20 +2,27 @@
...
-namespace Spiral\Console\Signature;
+namespace Spiral\Console\Configurator;
 
 use Symfony\Component\Console\Input\InputArgument;
 use Symfony\Component\Console\Input\InputOption;
 
+/**
+ * @internal
+ */
 final class Result
```
[the remark this answers is not in the document]
Comment (msmakouz) [reply src/Console/src/Configurator/Result.php]: Renamed

File: src/Console/src/PromptArguments.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+use Spiral\Attributes\ReaderInterface;
+use Spiral\Console\Attribute\Question;
+use Spiral\Console\Exception\ConsoleException;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+
+class PromptArguments
```
[the remark this answers is not in the document]
Comment (msmakouz) [reply src/Console/src/PromptArguments.php]: Added phpdoc and made it final

File: src/Console/src/Configurator/DefaultConfigurator.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+declare(strict_types=1);
+
+namespace Spiral\Console\Configurator;
+
+use Spiral\Console\Command;
+
+/**
+ * @internal
+ */
+final class DefaultConfigurator implements ConfiguratorInterface
```
[the remark this answers is not in the document]
Comment (msmakouz) [reply src/Console/src/Configurator/DefaultConfigurator.php]: Renamed to the **ConstantBasedConfigurator**

File: src/Console/src/Configurator/AttributeConfigurator.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace Spiral\Console\Configurator;
+
+use Spiral\Console\Command;
+use Spiral\Console\Configurator\Attribute\Parser;
+
+/**
+ * @internal
+ */
+final class AttributeConfigurator implements ConfiguratorInterface
```
[the remark this answers is not in the document]
Comment (msmakouz) [reply src/Console/src/Configurator/AttributeConfigurator.php]: Renamed

File: src/Console/src/Configurator/SignatureConfigurator.php [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace Spiral\Console\Configurator;
+
+use Spiral\Console\Command;
+use Spiral\Console\Configurator\Signature\Parser;
+
+/**
+ * @internal
+ */
+final class SignatureConfigurator implements ConfiguratorInterface
```
[the remark this answers is not in the document]
Comment (msmakouz) [reply src/Console/src/Configurator/SignatureConfigurator.php]: Renamed

**Review by butschster — APPROVED** [review]

**Review by butschster — APPROVED** [review]

**Pull request merged**